File: src/main/java/org/micoli/php/service/attributes/AttributeMapping.java
package org.micoli.php.service.attributes;

import com.jetbrains.php.lang.psi.elements.PhpAttribute;
import java.util.*;
import java.util.function.Function;
import org.jetbrains.annotations.NotNull;

public class AttributeMapping implements Cloneable {
    private final List<String> positionalParamsOrder;
    private final Map<String, Function<PhpAttribute.PhpAttributeArgument, String>> valueExtractors;
    private int positionalIndex = 0;

    public AttributeMapping(
            LinkedHashMap<String, Function<PhpAttribute.PhpAttributeArgument, String>> valueExtractors) {
        this.positionalParamsOrder = extractConstructorParameters(valueExtractors);
        this.valueExtractors = valueExtractors;
    }

    @Override
    public AttributeMapping clone() {
        try {
            return (AttributeMapping) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new AssertionError("Clone as failed", e);
        }
    }

    private static List<String> extractConstructorParameters(
            LinkedHashMap<String, Function<PhpAttribute.PhpAttributeArgument, String>> attributeClass) {
        if (attributeClass == null) {
            return Collections.emptyList();
        }

        return new ArrayList<>(attributeClass.keySet());
    }

    public String resolveParameterName(PhpAttribute.PhpAttributeArgument argument) {
        String name = argument.getName();
        if (name.isEmpty() && positionalIndex < positionalParamsOrder.size()) {
            name = positionalParamsOrder.get(positionalIndex++);
        }
        return name;
    }

    public Map<String, String> extractValues(@NotNull PhpAttribute attribute) {
        Map<String, String> extractedValues = new HashMap<>();

        Collection<PhpAttribute.PhpAttributeArgument> arguments = attribute.getArguments();
        for (PhpAttribute.PhpAttributeArgument argument : arguments) {
            String paramName = this.resolveParameterName(argument);
            String value = this.extractValue(paramName, argument);
            extractedValues.put(paramName, value);
        }
        return extractedValues;
    }

    public String extractValue(String paramName, PhpAttribute.PhpAttributeArgument argument) {
        Function<PhpAttribute.PhpAttributeArgument, String> callback = valueExtractors.getOrDefault(paramName, null);
        if (callback == null) {
            return null;
        }
        return callback.apply(argument);
    }
}

File: src/main/kotlin/org/micoli/php/service/attributes/AttributeMapping.kt
package org.micoli.php.service.attributes

import com.jetbrains.php.lang.psi.elements.PhpAttribute
import com.jetbrains.php.lang.psi.elements.PhpAttribute.PhpAttributeArgument
import java.util.function.Function
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.LinkedHashMap
import kotlin.collections.MutableCollection
import kotlin.collections.MutableList
import kotlin.collections.MutableMap

class AttributeMapping(
    private val valueExtractors: LinkedHashMap<String, Function<PhpAttributeArgument, String>>
) : Cloneable {
    private val positionalParamsOrder: MutableList<String> =
        extractConstructorParameters(valueExtractors)
    private var positionalIndex = 0

    public override fun clone(): AttributeMapping {
        try {
            return super.clone() as AttributeMapping
        } catch (e: CloneNotSupportedException) {
            throw AssertionError("Clone as failed", e)
        }
    }

    fun resolveParameterName(argument: PhpAttributeArgument): String {
        var name = argument.name
        if (name.isEmpty() && positionalIndex < positionalParamsOrder.size) {
            name = positionalParamsOrder[positionalIndex++]
        }
        return name
    }

    fun extractValues(attribute: PhpAttribute): MutableMap<String, String?> {
        val extractedValues: MutableMap<String, String?> = HashMap()

        val arguments: MutableCollection<PhpAttributeArgument> = attribute.arguments
        for (argument in arguments) {
            val paramName = resolveParameterName(argument)
            val value = extractValue(paramName, argument)
            extractedValues[paramName] = value
        }
        return extractedValues
    }

    fun extractValue(paramName: String, argument: PhpAttributeArgument): String? {
        val callback = valueExtractors.getOrDefault(paramName, null) ?: return null
        return callback.apply(argument)
    }

    companion object {
        private fun extractConstructorParameters(
            attributeClass: LinkedHashMap<String, Function<PhpAttributeArgument, String>>
        ): MutableList<String> {
            return ArrayList(attributeClass.keys)
        }
    }
}
